WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,767 @@
+   }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* get the group size */
+   usOverlayGroupSize = M_COMRV_GET_OVL_GROUP_SIZE(unToken);
+
+   /* if the data/function is not loaded we need to evict and load it */
+   if (pAddress == NULL)
+   {
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+      /* if the requested token is a multi-group token */
+      if (unToken.stFields.multiGroup != 0)
+      {
+         /* for now we take the first token in the list of tokens */
+         unToken = pOverlayMultiGroupTokensTable[unToken.stFields.overlayGroupID];
+         /* save the selected multi group entry */
+         usSelectedMultiGroupEntry = unToken.stFields.overlayGroupID;
+      }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+      /* disable ints */
+      // TODO: disable ints
+
+      /* get eviction candidates according to the requested pOverlayGroupSize */
+      ucNumOfEvictionCandidates = comrvGetEvictionCandidates(usOverlayGroupSize, ucEvictCandidateList);
+
+      ucEntryIndex = 0;
+      /* we need to handle heap fragmentation since we got more
+         than 1 eviction candidate */
+      if (ucNumOfEvictionCandidates > 1)
+      {
+         /* no handling to the last eviction candidate */
+         ucNumOfEvictionCandidates--;
+         while (ucEntryIndex < ucNumOfEvictionCandidates)
+         {
+            /* get the candidate entry */
+            pEntry = &stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]];
+            /* calc the source address */
+            pAddress = ((u08_t*)pEntry->pFixedEntryAddress + M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(pEntry->unToken));
+            /* perform code copy */
+            comrvMemcpyHook(pEntry->pFixedEntryAddress, pAddress,
+                  stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress - pAddress);
+
+            /* after code copy we need to align the entries structures */
+            for(ucIndex = ucEvictCandidateList[ucEntryIndex] ; ucIndex < ucEvictCandidateList[ucEntryIndex+1] ; ucIndex++)
+            {
+               pEntry = &stComrvCB.stOverlayCache[ucIndex + stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits];
+#ifdef D_COMRV_OVL_DATA_SUPPORT
+               /* an overlay data is present when handling defragmentation */
+               if (pEntry->unProperties.ucData)
+               {
+                  /* enable ints */
+                  // TODO: enable ints
+                  comrvNotificationHook(D_COMRV_OVL_DATA_DEFRAG_ERR, unToken.ucValue);
+               }
+#endif /* D_COMRV_OVL_DATA_SUPPORT */
+               stComrvCB.stOverlayCache[ucIndex].unProperties = pEntry->unProperties;
+               stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = pEntry->unToken.uiValue;
+#ifdef D_COMRV_EVICTION_LRU
+               stComrvCB.stOverlayCache[pEntry->ucPrevIndex].ucNextIndex = ucIndex;
+               stComrvCB.stOverlayCache[pEntry->ucNextIndex].ucPrevIndex = ucIndex;
```

Comment:
could be error on next and prev are the same

Reply:
done in 8701f9d